### PR TITLE
GetInspectorT

### PR DESCRIPTION
**Pull request by isovector**
Some of the ideas in #69

### CODE REVIEW

**Comment by isovector** [comment]
Look at the following program:

```haskell
{-# LANGUAGE TemplateHaskell #-}

import Control.Monad
import Polysemy
import Polysemy.State
import Polysemy.Error
import Polysemy.Internal.Tactics


doCB :: IO String -> IO ()
doCB m = m >>= putStrLn


data Callback m a where
  Callback :: m String -> Callback m ()

makeSem ''Callback


runCallback
    :: Member (Lift IO) r
    => (forall x. Sem r x -> IO x)
    -> Sem (Callback ': r) a
    -> Sem r a
runCallback lower = interpretH $ \case
  Callback cb -> do
    cb' <- runT cb
    ev <- runEvacuator <$> getEvacuatorT
    sendM $ doCB $ do
      v <- lower .@ runCallback $ cb'
      pure $ maybe ":(" id $ ev v
    getInitialStateT


main :: IO ()
main = do
  void . (runM .@ runCallback)
       . runState False
       $ do
    sendM $ putStrLn "hello world"
    callback $ show <$> get @Bool
    modify not
    callback $ show <$> get @Bool

  putStrLn "\n"

  void . (runM .@ runCallback)
       . runError @()
       $ do
    sendM $ putStrLn "hello world"
    callback $ throw ()
```

which prints

```
hello world
False
True


hello world
:(
```

**Comment by isovector** [comment]
@jkachmar this would let you do your retry thing

**Comment by isovector** [comment]
I don't see any reason not to merge this. The fancier stuff in #69 can be done in some sort of `polysemy-2` release.

TIME TO BIKESHED: 

What's a better name for this function: `f a -> Maybe a`, where `f` is some existential bastard?

**Comment by jhenahan** [comment]
I might bikeshed `hush` to `rightMay` or `rightMaybe` or `eitherMaybe`. All those names come up in Hoogle, but `hush` is the least obvious unless you go find the type signature.

As for `f a -> Maybe a`, the closest analog I can find is something like `gcopoint` (generic `copoint`), but that's not exactly informative. That said, taking a lesson from `Control.Functor.Copointed`'s `extract :: f a -> a` , maybe you could call it `extractMay` or `extractMaybe`.

**Comment by NickSeagull** [comment]
+1 to `extractMaybe`. In the end editors autocomplete the function names, so having long names that state what they do is awesome 😄

**Comment by isovector** [comment]
I'm keen on neither `hush` nor `extractMaybe`, since both have existing intuitions in the ecosystem, which don't apply here:

* `hush`: from the `errors` package. A natural transformation `Either e ~> Maybe`, _which silences the errors_. I agree that the type looks similar here, but the concern is the difference in intentions. We don't want to silence the errors, we just want to get our hands on the insides!
* `extractMaybe`: Extracting is the right idea; but the name is already taken by comonads. 

Something else to keep in mind is that the thing we're naming isn't pure. Its' type is more this:

```haskell
  :: Sem (WithTactics e f m r) (forall x. f x -> Maybe x)
```

**Comment by jhenahan** [comment]
Non-serious: `extractMaybe` -> `extractics`. 😆 

From the docs
```haskell
-- The @f@ type here is existential and corresponds to "whatever
-- state the other effects want to keep track of." @f@ is always
-- a 'Functor'.
```

Maybe something like `fromExternalState` or even `fromState`, though it's not necessarily `State`, so that could be misleading. `fromExternal` or `fromEnv`, maybe?

If you prefer punny names, `restate` might be cute.

**Comment by NickSeagull** [comment]
I like `fromExternal` and `restate` 😁

File: src/Polysemy/Internal/Tactics.hs [hunk 1]
```diff
@@ -96,6 +91,40 @@ getInitialStateT :: forall f m r e. Sem (WithTactics e f m r) (f ())
 getInitialStateT = send @(Tactics _ m (e ': r)) GetInitialState
 
 
+------------------------------------------------------------------------------
+-- | Get a natural transformation capable of potentially inspecting values
+-- inside of @f@. Binding the result of 'getInspectorT' produces a function that
+-- can sometimes peek inside values returned by 'bindT'.
```
Comment (NickSeagull) [inline src/Polysemy/Internal/Tactics.hs]: What are the cases when it cannot inspect the values? Perhaps it'd be nice to document them?

**Pull request merged**